src/Symfony/Component/Serializer/Normalizer/ChainNormalizer.php

```diff
@@ -0,0 +1,213 @@
+            return $normalizer->normalize($object, $format, $context);
+        }
+
+        if (null === $object || \is_scalar($object)) {
+            return $object;
+        }
+
+        if (\is_array($object) && !$object && ($context[Serializer::EMPTY_ARRAY_AS_OBJECT] ?? false)) {
+            return new \ArrayObject();
+        }
+
+        if (is_iterable($object)) {
+            if ($object instanceof \Countable && ($context[AbstractObjectNormalizer::PRESERVE_EMPTY_OBJECTS] ?? false) && !\count($object)) {
+                return new \ArrayObject();
+            }
+
+            $normalized = [];
+            foreach ($object as $key => $val) {
+                $normalized[$key] = $this->normalize($val, $format, $context);
+            }
+
+            return $normalized;
+        }
+
+        if (\is_object($object)) {
+            if ([] === $this->normalizers) {
+                throw new LogicException('You must register at least one normalizer to be able to normalize objects.');
+            }
+
+            throw new NotNormalizableValueException(sprintf('Could not normalize object of type "%s", no supporting normalizer found.', get_debug_type($object)));
+        }
+
+        throw new NotNormalizableValueException('An unexpected value could not be normalized: '.(!\is_resource($object) ? var_export($object, true) : sprintf('"%s" resource', get_resource_type($object))));
+    }
+
+    public function supportsNormalization(mixed $data, ?string $format = null, array $context = []): bool
+    {
+        return null !== $this->getNormalizer($data, $format, $context);
+    }
+
+    public function getSupportedTypes(?string $format): array
+    {
+        if (null === $format) {
+            $format = '__null__';
+        }
+
+        if (!isset($this->supportedCache[$format])) {
+            foreach ($this->normalizers as $normalizer) {
+                $this->supportedCache + $normalizer->getSupportedTypes($format);
+            }
+        }
+
+        return $this->supportedCache[$format];
+    }
+
+    /**
+     * Returns a matching normalizer.
+     *
+     * @param mixed       $data    Data to get the serializer for
+     * @param string|null $format  Format name, present to give the option to normalizers to act differently based on formats
+     * @param array       $context Options available to the normalizer
+     */
+    private function getNormalizer(mixed $data, ?string $format, array $context): ?NormalizerInterface
+    {
+        if (\is_object($data)) {
+            $type = $data::class;
+            $genericType = 'object';
+        } else {
+            $type = 'native-'.\gettype($data);
+            $genericType = '*';
+        }
+
+        if (!isset($this->normalizerCache[$format][$type])) {
+            $this->normalizerCache[$format][$type] = [];
+
+            foreach ($this->normalizers as $k => $normalizer) {
+                if (!$normalizer instanceof NormalizerInterface) {
```

Comment:
this does not make sense to me. Once we separate the lists, we should not have anything else than a NormalizerInterface here